File: src/livekit_agent_simulator/script_parse.py
from __future__ import annotations

from typing import Any

from .script import (
    SUPPORTED_ACTIONS,
    SUPPORTED_TRIGGERS,
    ScriptStep,
    ScriptVerifySpec,
    normalize_interrupt_class,
)


def _parse_step_gain(raw: dict[str, Any], path_label: str, step_id: str) -> float:
    key = "gain"
    if "gain" not in raw and "volume" in raw:
        key = "volume"
    if key not in raw:
        return 1.0
    try:
        gain = float(raw[key])
    except (TypeError, ValueError) as e:
        raise ValueError(
            f"{path_label}: Script step {step_id!r}: {key} must be a number"
        ) from e
    if not 0.0 <= gain <= 1.0:
        raise ValueError(
            f"{path_label}: Script step {step_id!r}: {key} must be between 0.0 and 1.0"
        )
    return gain


def parse_script_verify(raw: Any) -> ScriptVerifySpec | None:
    if raw is None:
        return None
    if not isinstance(raw, dict):
        raise ValueError("Script.spec.verify must be an object")
    plugins_raw = raw.get("plugins")
    if plugins_raw is None and raw.get("plugin"):
        plugins_raw = [raw.get("plugin")]
    plugins: tuple[str, ...] = ()
    if plugins_raw is not None:
        if not isinstance(plugins_raw, list):
            raise ValueError("Script.spec.verify.plugins must be an array of plugin names")
        plugins = tuple(str(p).strip() for p in plugins_raw if str(p).strip())

    options_raw = raw.get("plugin_options")
    if options_raw is not None and not isinstance(options_raw, dict):
        raise ValueError("Script.spec.verify.plugin_options must be an object")
    plugin_options = dict(options_raw) if isinstance(options_raw, dict) else {}

    return ScriptVerifySpec(
        require_during_agent_speech=bool(raw.get("require_during_agent_speech", True)),
        min_agent_finals_after_first_cue=int(raw.get("min_agent_finals_after_first_cue", 0)),
        min_user_finals_after_first_cue=int(raw.get("min_user_finals_after_first_cue", 0)),
        min_interruptions=int(raw["min_interruptions"])
        if raw.get("min_interruptions") is not None
        else None,
        max_interruptions=int(raw["max_interruptions"])
        if raw.get("max_interruptions") is not None
        else None,
        min_agent_finals_after_silence=int(raw.get("min_agent_finals_after_silence", 0)),
        min_agent_finals_after_barge_in=int(raw.get("min_agent_finals_after_barge_in", 0)),
        plugins=plugins,
        plugin_options=plugin_options,
    )


def parse_script_steps(spec: dict[str, Any], path_label: str) -> list[ScriptStep]:
    raw_steps = spec.get("steps")
    if raw_steps is None:
        return []
    if not isinstance(raw_steps, list):
        raise ValueError(f"{path_label}: Script.spec.steps must be an array")

    steps: list[ScriptStep] = []
    for i, raw in enumerate(raw_steps):
        if not isinstance(raw, dict):
            raise ValueError(f"{path_label}: Script.spec.steps[{i}] must be an object")
        step_id = str(raw.get("id") or raw.get("label") or f"step-{i}")
        trigger = str(raw.get("trigger", "agent_speaking"))
        if trigger not in SUPPORTED_TRIGGERS:
            raise ValueError(
                f"{path_label}: Script step {step_id!r}: unsupported trigger {trigger!r} "
                f"(supported: {sorted(SUPPORTED_TRIGGERS)})"
            )
        action = str(raw.get("action", "speak"))
        if action not in SUPPORTED_ACTIONS:
            raise ValueError(
                f"{path_label}: Script step {step_id!r}: action must be speak|wait"
            )
        say = raw.get("say") or raw.get("text") or ""
        if action == "speak" and not str(say).strip():
            raise ValueError(f"{path_label}: Script step {step_id!r}: say/text required when action=speak")
        delivery = str(raw.get("delivery", "gemini_text"))
        if delivery not in ("gemini_text", "room_pcm"):
            raise ValueError(
                f"{path_label}: Script step {step_id!r}: delivery must be gemini_text or room_pcm"
            )
        asset = raw.get("asset")
        if action == "speak" and delivery == "room_pcm" and not asset:
            raise ValueError(
                f"{path_label}: Script step {step_id!r}: room_pcm delivery requires asset (WAV path)"
            )
        # Barge-in convenience: short delay while agent speaking
        delay_ms = int(raw.get("delay_ms", 800))
        min_agent = int(raw.get("min_agent_active_ms", 400))
        barge_in = bool(raw.get("barge_in") or raw.get("interrupt"))
        if barge_in:
            delay_ms = int(raw.get("delay_ms", 250))
            min_agent = int(raw.get("min_agent_active_ms", 200))
            trigger = "agent_speaking"
            action = "speak"
        # Default: blip on text barge; off when room_pcm (asset is the cut-in audio).
        if "with_blip" in raw:
            with_blip = bool(raw.get("with_blip"))
        else:
            with_blip = barge_in and delivery != "room_pcm"
        gain = _parse_step_gain(raw, path_label, step_id)
        loop = bool(raw.get("loop") or raw.get("repeat") is True)
        if "repeat" in raw and not isinstance(raw.get("repeat"), bool):
            # repeat as count is not supported; only bool alias for loop
            if raw.get("repeat") not in (None, 0, 1, "0", "1"):
                raise ValueError(
                    f"{path_label}: Script step {step_id!r}: use loop=true for continuous "
                    f"ambient beds (repeat count is not supported)"
                )
        if loop and delivery != "room_pcm":
            raise ValueError(
                f"{path_label}: Script step {step_id!r}: loop requires delivery=room_pcm"
            )
        if loop and action != "speak":
            raise ValueError(
                f"{path_label}: Script step {step_id!r}: loop only applies to speak + room_pcm"
            )
        asset_s = str(asset).strip() if asset else None
        if loop and asset_s:
            name = asset_s.lower()
            if name.startswith("builtin:"):
                name = name[len("builtin:") :]
            if name.startswith("@"):
                name = name[1:]
            if name.startswith("voice."):
                raise ValueError(
                    f"{path_label}: Script step {step_id!r}: loop is for noise/ambient beds, "
                    f"not voice.* speech assets"
                )
        try:
            interrupt_class = normalize_interrupt_class(
                raw.get("class") or raw.get("interrupt_class"),
                barge_in=barge_in,
            )
        except ValueError as e:
            raise ValueError(f"{path_label}: Script step {step_id!r}: {e}") from e

        overlay_raw = raw.get("overlay") or raw.get("speech_role")
        overlay: str | None = None
        if overlay_raw is not None and str(overlay_raw).strip():
            overlay = str(overlay_raw).strip().lower().replace("-", "_")
            if overlay in ("forced_line", "forced", "say"):
                overlay = "line"
            if overlay not in ("fixture", "line"):
                raise ValueError(
                    f"{path_label}: Script step {step_id!r}: overlay must be "
                    f"fixture|line (got {overlay_raw!r})"
                )

        steps.append(
            ScriptStep(
                id=step_id,
                trigger=trigger,
                delay_ms=delay_ms,
                say=str(say).strip(),
                label=str(raw.get("label") or step_id),
                once=bool(raw.get("once", True)),
                min_agent_active_ms=min_agent,
                delivery=delivery,
                asset=asset_s,
                silence_after_cue_ms=int(raw.get("silence_after_cue_ms", 0)),
                action=action,
                require_agent_spoke_first=bool(raw.get("require_agent_spoke_first", True)),
                require_agent_reply_this_turn=bool(
                    raw.get("require_agent_reply_this_turn", True)
                ),
                defer_on_open_question=bool(raw.get("defer_on_open_question", True)),
                open_question_idle_ms=int(raw.get("open_question_idle_ms", 20000)),
                barge_in=barge_in,
                with_blip=with_blip,
                gain=gain,
                interrupt_class=interrupt_class,
                overlay=overlay,
                loop=loop,
            )
        )
    return steps

File: src/livekit_agent_simulator_rust/crates/lks-core/tests/script_parse.rs
//! Golden tests for script step/verify parsing (P1) — vs script_parse.py.
use lks_core::script::parse::{parse_script_steps, parse_script_verify};
use lks_core::script::{counts_for_recovery_barge, effective_overlay, normalize_interrupt_class};
use serde_json::json;

#[test]
fn parse_simple_speak_step() {
    let spec = json!({
        "steps": [{
            "id": "greet",
            "trigger": "agent_speaking",
            "delay_ms": 500,
            "say": "Hello!",
            "action": "speak"
        }]
    });
    let steps = parse_script_steps(spec.as_object().unwrap(), "test").expect("parse");
    assert_eq!(steps.len(), 1);
    let s = &steps[0];
    assert_eq!(s.id, "greet");
    assert_eq!(s.trigger, "agent_speaking");
    assert_eq!(s.delay_ms, 500);
    assert_eq!(s.say, "Hello!");
    assert_eq!(s.action, "speak");
    assert_eq!(s.delivery, "gemini_text");
    assert!(s.once);
    assert_eq!(s.min_agent_active_ms, 400);
    assert_eq!(s.gain, 1.0);
    assert_eq!(s.interrupt_class, None);
}

#[test]
fn parse_barge_in_forces_defaults() {
    let spec = json!({
        "steps": [{
            "id": "barge",
            "say": "Actually...",
            "barge_in": true,
            "interrupt_class": "correction"
        }]
    });
    let steps = parse_script_steps(spec.as_object().unwrap(), "test").expect("parse");
    let s = &steps[0];
    // barge_in forces trigger=agent_speaking, action=speak, shorter defaults
    assert_eq!(s.trigger, "agent_speaking");
    assert_eq!(s.action, "speak");
    assert_eq!(s.delay_ms, 250);
    assert_eq!(s.min_agent_active_ms, 200);
    assert!(s.barge_in);
    assert_eq!(s.interrupt_class.as_deref(), Some("correction"));
    assert!(s.with_blip, "barge + gemini_text → blip on");
}

#[test]
fn parse_dtmf_step() {
    let spec = json!({
        "steps": [{
            "id": "dial",
            "action": "dtmf",
            "digits": "123#"
        }]
    });
    let steps = parse_script_steps(spec.as_object().unwrap(), "test").expect("parse");
    let s = &steps[0];
    assert_eq!(s.action, "dtmf");
    assert_eq!(s.digits, "123#");
    assert_eq!(s.say, "[DTMF: 123#]");
    assert_eq!(s.trigger, "time", "dtmf without silence/time → time");
}

#[test]
fn parse_dtmf_invalid_digit() {
    let spec = json!({
        "steps": [{
            "id": "bad",
            "action": "dtmf",
            "digits": "12x"
        }]
    });
    let err =
        parse_script_steps(spec.as_object().unwrap(), "test").expect_err("invalid digit fails");
    assert!(err.contains("digits can only contain 0-9*#w"), "got: {err}");
}

#[test]
fn parse_room_pcm_requires_asset() {
    let spec = json!({
        "steps": [{
            "id": "noise",
            "delivery": "room_pcm",
            "say": "x"
        }]
    });
    let err = parse_script_steps(spec.as_object().unwrap(), "test").expect_err("no asset fails");
    assert!(
        err.contains("room_pcm delivery requires asset"),
        "got: {err}"
    );
}

#[test]
fn parse_loop_requires_room_pcm() {
    let spec = json!({
        "steps": [{
            "id": "amb",
            "loop": true,
            "say": "x"
        }]
    });
    let err =
        parse_script_steps(spec.as_object().unwrap(), "test").expect_err("loop needs room_pcm");
    assert!(
        err.contains("loop requires delivery=room_pcm"),
        "got: {err}"
    );
}

#[test]
fn parse_say_required_for_speak() {
    let spec = json!({
        "steps": [{"id": "no-say", "action": "speak"}]
    });
    let err = parse_script_steps(spec.as_object().unwrap(), "test").expect_err("no say fails");
    assert!(
        err.contains("say/text required when action=speak"),
        "got: {err}"
    );
}

#[test]
fn parse_unsupported_trigger() {
    let spec = json!({
        "steps": [{"id": "bad", "trigger": "never", "say": "x"}]
    });
    let err = parse_script_steps(spec.as_object().unwrap(), "test").expect_err("bad trigger fails");
    assert!(err.contains("unsupported trigger"), "got: {err}");
}

#[test]
fn normalize_interrupt_class_aliases() {
    // barge/correct → correction; ack → backchannel; noise; dtmf; escalate
    assert_eq!(
        normalize_interrupt_class(Some(&json!("barge")), false, "correction").unwrap(),
        Some("correction".into())
    );
    assert_eq!(
        normalize_interrupt_class(Some(&json!("ack")), false, "correction").unwrap(),
        Some("backchannel".into())
    );
    assert_eq!(
        normalize_interrupt_class(Some(&json!("false_positive")), false, "correction").unwrap(),
        Some("noise".into())
    );
    assert_eq!(
        normalize_interrupt_class(Some(&json!("handoff")), false, "correction").unwrap(),
        Some("escalate".into())
    );
    // barge_in with no class → correction
    assert_eq!(
        normalize_interrupt_class(None, true, "correction").unwrap(),
        Some("correction".into())
    );
    // no barge, no class → None
    assert_eq!(
        normalize_interrupt_class(None, false, "correction").unwrap(),
        None
    );
}

#[test]
fn counts_for_recovery_barge_cases() {
    assert!(counts_for_recovery_barge(true, Some("correction")));
    assert!(counts_for_recovery_barge(true, Some("escalate")));
    assert!(!counts_for_recovery_barge(true, Some("noise")));
    assert!(!counts_for_recovery_barge(false, Some("correction")));
    // barge with None class → Python defaults to "correction" → counts
    assert!(
        counts_for_recovery_barge(true, None),
        "None class → correction → counts"
    );
}

#[test]
fn effective_overlay_cases() {
    let base = |id: &str| lks_core::script::ScriptStep {
        id: id.into(),
        trigger: "agent_speaking".into(),
        delay_ms: 400,
        say: "hi".into(),
        label: id.into(),
        once: true,
        min_agent_active_ms: 400,
        delivery: "gemini_text".into(),
        asset: None,
        silence_after_cue_ms: 0,
        action: "speak".into(),
        mute_persona: None,
        digits: "".into(),
        r#loop: false,
        require_agent_spoke_first: true,
        require_agent_reply_this_turn: true,
        defer_on_open_question: true,
        open_question_idle_ms: 20_000,
        barge_in: false,
        with_blip: false,
        gain: 1.0,
        interrupt_class: None,
        overlay: None,
    };
    // explicit overlay wins
    let mut s = base("a");
    s.overlay = Some("line".into());
    assert_eq!(effective_overlay(&s), "line");
    // barge_in → fixture
    let mut s = base("b");
    s.barge_in = true;
    assert_eq!(effective_overlay(&s), "fixture");
    // room_pcm → fixture
    let mut s = base("c");
    s.delivery = "room_pcm".into();
    assert_eq!(effective_overlay(&s), "fixture");
    // speak with say → line
    assert_eq!(effective_overlay(&base("d")), "line");
}

#[test]
fn parse_script_verify_spec() {
    let v = json!({
        "require_during_agent_speech": false,
        "min_agent_finals_after_first_cue": 1,
        "min_agent_finals_after_barge_in": 2,
        "plugins": ["a", "b"]
    });
    let vs = parse_script_verify(&v).expect("verify").expect("some");
    assert!(!vs.require_during_agent_speech);
    assert_eq!(vs.min_agent_finals_after_first_cue, 1);
    assert_eq!(vs.min_agent_finals_after_barge_in, 2);
    assert_eq!(vs.plugins, vec!["a", "b"]);
    // None when absent
    assert!(parse_script_verify(&json!(null)).unwrap().is_none());
}
